Don't assume memory layout of std::net::SocketAddr

faern opened the pull request:
Fixes #105 

This crate assumes the memory layout of `std::net::SocketAddr`, something it should not do. The layout might be correct now, but it's not part of the contract `std` makes to the outside world. Having commonly used crates like this one depend on unspecified things like that makes it harder for `std` to evolve.

This PR changes that by correctly converting between the `std` representation and the system representation instead of just doing pointer casting.

Same kind of fix as was recently done in https://github.com/rust-lang/socket2-rs/pull/120 and https://github.com/tokio-rs/mio/pull/1388

faern committed:
Safely convert SocketAddr -> system representation

faern committed:
Use transmute instead of u32::from_ne_bytes to support older Rust

faern committed:
Initialize in6_addr in a way that works for older Rust

faern commented:
All of these PRs look a bit different. Due to how the libraries are structured, but also because they have different minimum supported Rust versions etc. I'm assuming we have to keep the compatibility with Rust 1.21.0 for this fix. We want as many people as possible to be able to upgrade as smoothly as possible to get rid of the possible unsoundness this code had.

faern commented:
Ok. Done. Ready for review I believe. Ping @pfmooney ?

For some context, see this PR: https://github.com/rust-lang/rust/pull/78802. And possibly this internals thread: https://internals.rust-lang.org/t/why-are-socketaddrv4-socketaddrv6-based-on-low-level-sockaddr-in-6/13321

Thomasdezeeuw reviewed (approved):
The same changes were applied to socket2 and Mio.

faern commented:
I realize this PR still has a lot of transmutes that I aimed to get rid of. The other PRs have better solutions to this. I will fix this later today.

Thomasdezeeuw commented:
> I realize this PR still has a lot of transmutes that I aimed to get rid of. The other PRs have better solutions to this. I will fix this later today.

I was looking at that but the `be_bytes` functions are Rust 1.32.0+, which is higher than the minimum supported version. But you can use `swap_bytes`.

faern commented:
Oh yeah. That's why I moved over to `transmute`. I forgot. Too many very similar PRs this weekend :sweat_smile:. In another PR we temporarily had transmutes for other reasons that we got rid of. No this PR is good to go IMO.

I don't want any byte swapping. I just want to convert `u32->[u8;4]` and back without any swapping. I could also just do a pointer cast instead of a transmute. But I'll let the reviewer of this crate decide on that. The current implementation should work and be good enough IMO.

faern commented:
I have verified that both `miow` and `mio` builds and completes their test suites when based on this version of `net2`. Both on latest stable Rust and on the Rust toolchain built from https://github.com/rust-lang/rust/pull/78802. So I'm pretty confident this is correctly implemented.

@pfmooney Would it be possible to have this reviewed (and published) :see_no_evil:? This crate is at the bottom of a pretty large dependency chain doing this invalid memory assumption. So things kind of have to start here.

pfmooney commented on src/socket.rs:
```diff
@@ -84,13 +84,101 @@ impl ::IntoInner for Socket {
-        SocketAddr::V6(ref a) => {
-            (a as *const _ as *const _, mem::size_of_val(a) as c::socklen_t)
+        SocketAddr::V6(addr) => {
+            #[cfg(unix)]
+            let sin6_addr = {
+                let mut sin6_addr = unsafe { mem::zeroed::<c::in6_addr>() };
+                sin6_addr.s6_addr = addr.ip().octets();
+                sin6_addr
+            };
+            #[cfg(windows)]
+            let sin6_addr = unsafe {
+                let mut u = mem::zeroed::<c::in6_addr_u>();
+                *u.Byte_mut() = addr.ip().octets();
+                c::IN6_ADDR { u }
+            };
+            #[cfg(windows)]
+            let u = unsafe {
+                let mut u = mem::zeroed::<c::SOCKADDR_IN6_LH_u>();
+                *u.sin6_scope_id_mut() = addr.scope_id();
+                u
+            };
```
What are your thoughts about splitting `addr2raw` into separate `cfg(windows)` and `cfg(unix)` variants, rather than interleaving like this?  It might be a little easier to follow, event though the `unix` version will still include the various specialization bits.

faern replied:
Done. I also splitted the v4 and v6 versions since the function was so long. I think this makes sense and puts most of the code at way lower indentation.

pfmooney commented on tests/all.rs:
```diff
@@ -16,10 +16,11 @@ macro_rules! t {
 fn smoke_build_listener() {
     let b = t!(TcpBuilder::new_v4());
-    t!(b.bind("127.0.0.1:0"));
+    t!(b.bind("127.0.0.1:6543"));
```
If we're going to hard-code the port, what about picking something a little higher (16384-32768) and not covered by an `/etc/services` listing?  It would give us a better chance of avoiding a clash with an existing thing, should the tests be run on a system with other running services.

faern replied:
Now when I test it it's actually quite likely to run into "Address already in use". Not because something else happens to listen there. But when running the tests multiple times and the socket does not get cleaned up fast enough.. So what do you think. Should we skip testing this?

pfmooney replied:
Failing on repeat runs (due to lingering sockets) seems less than ideal.  Skipping it probably makes sense for now.

faern replied:
I think we're fine without it. Nothing tested this was correct before either. And I'm pretty confident it's correct if I can make it pass the `mio` and `tokio` test suites.

pfmooney commented on tests/all.rs:
```diff
@@ -38,8 +39,9 @@ fn smoke_build_listener() {
 fn smoke_build_listener_v6() {
     let b = t!(TcpBuilder::new_v6());
-    t!(b.bind("::1:0"));
+    t!(b.bind("::1:1234"));
```
Same as above, Re: port selection

pfmooney commented on src/socket.rs:
```diff
@@ -84,13 +84,101 @@ impl ::IntoInner for Socket {
+/// A type with the same memory layout as `c::sockaddr`. Used in converting Rust level
+/// SocketAddr* types into their system representation. The benefit of this specific
+/// type over using `c::sockaddr_storage` is that this type is exactly as large as it
+/// needs to be and not a lot larger. And it can be initialized cleaner from Rust.
```
nit: Skip the "And it can be initialized cleaner from Rust." sentence fragment

pfmooney reviewed (commented):
Thanks for putting this together, and please pardon the review delay.

faern committed:
Remove unwanted part of docs

faern committed:
Split addr2raw up into separate platform functions

pfmooney commented on src/socket.rs:
```diff
@@ -84,13 +84,101 @@ impl ::IntoInner for Socket {
+            // So just transmuting the octects to the `u32` representation works.
+            #[cfg(unix)]
+            let sin_addr = c::in_addr {
+                s_addr: unsafe { mem::transmute::<_, u32>(addr.ip().octets()) },
```
Will the alignment requirements of `u32` be fulfilled by the array emitted from `ip().octets()`?  If that isn't guaranteed, is it really that expensive to do with safe rust?

faern replied:
I'm not 100% sure if this is true for all platforms, but I think it should be. The alignment of `Ipv4Addr` is 4, which is the same as `u32`. So the alignment of the `[u8; 4]` should be correct unless I'm mistaken.

We could solve it safely by using the built in `Into<u32>` etc. And the compiler might be able to optimize it away, but if not it will do a byte swap twice.

pfmooney replied:
I think avoiding the `unsafe` is worth the minuscule cost of a double byte-swap (especially considering it might be optimized away).

faern replied:
Correctness is of course way more important than performance. So if we can't say for sure that this is correct I agree we should change to a safer solution.

faern replied:
Fixed. Less `unsafe` is always nice!

faern replied:
Oops. I realize `Ipv4Addr::octets` of course return an owned `[u8; 4]` and not a reference. As such, the IP data is copied onto the stack and the alignment is just 1, not 4. Good we changed it.

faern committed:
Fix references

faern committed:
Replace transmutes with safe convertion

faern commented:
@pfmooney There were a lot of pushes in a row there. But now I have addressed all the feedback and I have tested what's currently in this PR under `miow` and `mio 0.6` successfully.

pfmooney commented:
> @pfmooney There were a lot of pushes in a row there. But now I have addressed all the feedback and I have tested what's currently in this PR under `miow` and `mio 0.6` successfully.

Great.  I'll run it through a few smoke tests of my own while waiting for @Thomasdezeeuw to approve of the updated change.

Thomasdezeeuw reviewed (approved):
LGTM. @faern I'm amusing you've tried this branch with Mio and all tests passes?

faern commented:
> @faern I'm amusing you've tried this branch with Mio and all tests passes?

Yes. On Windows 10 and Linux only though. But all tests pass. Both on a standard build of Rust and on the custom build where I change the memory layout of `SocketAddr`. So with this PR it indeed seem to not rely on the memory layout :)

faern commented:
@pfmooney ping?

pfmooney commented:
Merged as 49b43f277afb1caf5104fcbe02bef581f7444686.  Thanks

pfmooney commented:
...and published as [0.2.36](https://crates.io/crates/net2/0.2.36).

faern commented:
Yay! Thank you :tada:
